static/panes/compiler.js

```diff
@@ -130,33 +129,77 @@ function Compiler(hub, container, state) {
 
     this.fontScale = new FontScale(this.domRoot, state, this.outputEditor);
 
-    this.compilerPicker.selectize({
+    var self = this;
+    var optgroups = this.compilerService.getGroupsInUse(this.currentLangId);
+    optgroups.unshift({value:'favs',label:'Favorites'});
+
+    this.compilerSelectizer = new TomSelect(this.compilerPicker[0],{
         sortField: this.compilerService.getSelectizerOrder(),
         valueField: 'id',
         labelField: 'name',
         searchField: ['name'],
         optgroupField: 'group',
-        optgroups: this.compilerService.getGroupsInUse(this.currentLangId),
+        optgroups: optgroups,
         lockOptgroupOrder: true,
         options: _.filter(this.getCurrentLangCompilers(), function (e) {
             return !e.hidden || e.id === state.compiler;
         }),
         items: this.compiler ? [this.compiler.id] : [],
         dropdownParent: 'body',
         closeAfterSelect: true,
-    }).on('change', _.bind(function (e) {
-        var val = $(e.target).val();
-        if (val) {
-            ga.proxy('send', {
-                hitType: 'event',
-                eventCategory: 'SelectCompiler',
-                eventAction: val,
-            });
-            this.onCompilerChange(val);
+        plugins:['dropdown_input'],
+        onChange: function (val){
+            if (val) {
+                ga.proxy('send', {
+                    hitType: 'event',
+                    eventCategory: 'SelectCompiler',
+                    eventAction: val,
+                });
+                self.onCompilerChange(val);
+            }
+        },
+        duplicates: true,
+        render:{
+            option:function (data,escape){
+                return '<div class="d-flex"><div>'
+					+ escape(data.name)
+					+ '</div><div class="ml-auto toggle-fav"><i class="fas fa-star"></i></div></div>';
+            },
+        },
+    });
+
+    $(this.compilerSelectizer.dropdown_content).on('click','.toggle-fav',function (evt){
+        evt.preventDefault();
+        evt.stopPropagation();
+
+        var option_el        = this.closest('.option');
+        var value            = option_el.dataset.value;
+        var data            = self.compilerSelectizer.options[value];
+        var remove_fav        = false;
+
+        data.group = Array.isArray(data.group) ? data.group : [data.group];
+
+        if( option_el.classList.contains('fav') ){
+            remove_fav = true;
+            var index = data.group.indexOf('favs');
+            data.group.splice(index, 1);
+
+        }else{
+            data.group.push('favs');
+            option_el.classList.add('fav');
         }
-    }, this));
 
-    this.compilerSelectizer = this.compilerPicker[0].selectize;
+        self.compilerSelectizer.refreshOptions(false);
+
+
+        // when removing, there are two options with the fav class until after refreshOptions()
```

Comment:
I don't understand this comment - can you elaborate?

Reply:
This relates to the duplicates you asked about above. When "armv7-a clang 11.0.1" is added to a user's favorites, there will be two options for that compiler. Both of those options have a star that, when clicked, will remove the compiler from the favorites. If a user clicks the star under the "favorites" heading, the option under the "ARM 32-bit clang" heading still needs to be updated. So, instead of figuring out which option the user clicked on, the code just waits till after refreshOptions() finishes rebuilding the dropdown options, at which point there will be only one option for "armv7-a clang 11.0.1".

Sorry, long winded, but hopefully it's clear.

Reply:
That does make sense, thank you.